src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs

```diff
@@ -1340,6 +1341,94 @@ public bool WaitForExit(int milliseconds)
+        /// for the <paramref name="cancellationToken"/> to be canceled.
+        /// </summary>
+        /// <remarks>
+        /// Calling this method will set <see cref="EnableRaisingEvents"/> to true.
+        /// </remarks>
+        public async Task WaitForExitAsync(CancellationToken cancellationToken = default)
+        {
+            // Because the process has already started by the time this method is called,
+            // we're in a race against the process to set up our exit handlers before the process
+            // exits. As a result, there are several different flows that must be handled:
+            //
+            // CASE 1: WE ENABLE EVENTS
+            // This is the "happy path". In this case we enable events.
+            //
+            // CASE 1.1: PROCESS EXITS OR IS CANCELED AFTER REGISTERING HANDLER
+            // This case continues the "happy path". The process exits or waiting is canceled after
+            // registering the handler and no special cases are needed.
+            //
+            // CASE 1.2: PROCESS EXITS BEFORE REGISTERING HANDLER
+            // It's possible that the process can exit after we enable events but before we reigster
+            // the handler. In that case we must check for exit after registering the handler.
+            //
+            //
+            // CASE 2: PROCESS EXITS BEFORE ENABLING EVENTS
+            // The process may exit before we attempt to enable events. In that case EnableRaisingEvents
+            // will throw an exception like this:
+            //     System.InvalidOperationException : Cannot process request because the process (42) has exited.
+            // In this case we catch the InvalidOperationException. If the process has exited, our work
+            // is done and we return. If for any reason (now or in the future) enabling events fails
+            // and the process has not exited, bubble the exception up to the user.
+            //
+            //
+            // CASE 3: USER ALREADY ENABLED EVENTS
+            // In this case the user has already enabled raising events. Re-enabling events is a no-op
+            // as the value hasn't changed. However, no-op also means that if the process has already
+            // exited, EnableRaisingEvents won't throw an exception.
+            //
+            // CASE 3.1: PROCESS EXITS OR IS CANCELED AFTER REGISTERING HANDLER
+            // (See CASE 1.1)
+            //
+            // CASE 3.2: PROCESS EXITS BEFORE REGISTERING HANDLER
+            // (See CASE 1.2)
+
+            if (!Associated) { throw new InvalidOperationException(SR.NoAssociatedProcess); }
+
+            try
+            {
+                // CASE 1: We enable events
+                // CASE 2: Process exits before enabling events (and throws an exception)
+                // CASE 3: User already enabled events (no-op)
+                EnableRaisingEvents = true;
+            }
+            catch (InvalidOperationException)
```

Comment:
EnableRaisingEvents is throwing an exception if the associated process already exited?  Can you share a repro of that?

Reply:
Sure! I originally hit this in the unit test `System.Diagnostics.Tests.ProcessWaitingTests.SingleProcess_CopiesShareExitAsyncInformation`.

Here's a console app that repros the issue so you don't need to be inside an xunit project:

```csharp
using System;
using System.Diagnostics;
using System.Threading;

class Program
{
    static void Main(string[] args)
    {
        var timeout = TimeSpan.FromSeconds(5);

        var p = new Process
        {
            StartInfo = new ProcessStartInfo
            {
                FileName = "cmd.exe",
                Arguments = $"/c timeout /t {timeout.TotalSeconds} /nobreak",
                UseShellExecute = false,
            }
        };

        Console.WriteLine("Staring main process...");
        p.Start();
        Console.WriteLine($"PID = {p.Id}");

        Console.WriteLine("Creating 'copy' of process by PID");
        var copy = Process.GetProcessById(p.Id);

        Console.WriteLine("Waiting for main process to exit...");
        Thread.Sleep(timeout * 2);

        Console.WriteLine("Enabling events on main process");
        p.EnableRaisingEvents = true;

        Console.WriteLine("Enabling events on 'copy'");
        copy.EnableRaisingEvents = true;

        Console.WriteLine("Done!");
    }
}
```

And here's the output:

```
Staring main process...
PID = 37344
Creating 'copy' of process by PID
Waiting for main process to exit...

Waiting for 0 seconds, press CTRL+C to quit ...
Enabling events on main process
Enabling events on 'copy'
Unhandled exception. System.InvalidOperationException: Cannot process request because the process (37344) has exited.
   at System.Diagnostics.Process.GetProcessHandle(Int32 access, Boolean throwIfExited)
   at System.Diagnostics.Process.GetOrOpenProcessHandle()
   at System.Diagnostics.Process.EnsureWatchingForExit()
   at System.Diagnostics.Process.set_EnableRaisingEvents(Boolean value)
   at Program.Main(String[] args) in C:\Users\mattkot\Desktop\temp\ConsoleApp1\Program.cs:line 35
```

This snippet is from a netcoreapp3.1 running on Windows 10 Version	10.0.18363 Build 18363

Reply:
I see. Thanks.  EnableRaisingEvents is eventually calling GetProcessHandle with throwIfExited:true.

For now, your workaround is fine.  Ideally though we'd avoid that exception in the first place rather than letting it be thrown and eating it.

Reply:
Thanks Stephen! To be very explicit about next steps, I see a few paths forward, which one do you prefer?

1. Merge PR as-is and file an issue about the `EnableRaisingEvents` behavior
2. Refactor the `EnableRaisingEvents` code path to avoid throwing the exception for this internal call-site, but keep the public behavior the same
3. Make a separate PR to change the behavior for `EnableRaisingEvents`, then rebase this PR on top of that one

My least favorite option is 2 because it hints that there's a flaw in the API and doesn't solve the issue for other devs. That said, the behavior of `EnableRaisingEvents` is a bit inconsistent either way depending on how you look at it. For instance, `MaxWorkingSet` and `Modules` _do_ throw if called after the process exits, so it's semi-reasonable to expect properties to throw on exited processes.

Let me know what path you prefer. Thanks!

Reply:
Eventually I think we should do 2. For now an issue is fine.

Reply:
Filed #1785.